lms/djangoapps/mobile_api/course_info/utils.py

```diff
@@ -0,0 +1,43 @@
+"""
+Common utils for the `course_info` API.
+"""
+
+import logging
+from typing import List, Optional, Union
+
+from django.core.cache import cache
+
+from lms.djangoapps.courseware.access import has_access
+from lms.djangoapps.grades.api import CourseGradeFactory
+from openedx.core.djangoapps.content.block_structure.api import get_block_structure_manager
+
+log = logging.getLogger(__name__)
+
+
+def calculate_progress(
```

Comment:
Same as my comment on `get_course_assignments`, this is core business logic that needs to be implemented in core APIs for consistency with the Web views.

Reply:
And this `calculate_progress ` function I also move to `lms/djangoapps/courseware/courses.py`. Everything works correctly and I'm going to push these changes

Reply:
Sounds good. Let me know when you're ready for another review pass.

Reply:
@kdmccormick the comments were addressed, It's ready for the second round now

Reply:
@KyryloKireiev I see that you moved one definition of `calculate_progress` to lms/djangoapps/courseware/courses.py, thank you. However, there is this second definition of `calculate_progress` below, still in the mobile API.

Firstly, I think this function should be called something else, as it returns subsection grades, not progress.

Secondly, and more importantly, I am concerned whether this implementation matches the other ways in which we show a user's subsection grades in edx-platform. For example, I assume that the gradebook and the progress page on the Learning MFE both show subsection grades. Does those implementations perform the same steps that we do here (load the block structure, cache it, calculate grades, and then re-calculate from visible-only) ?  If so, why not use a shared API function? If not, why?

cc @jawad-khan

Reply:
1. @kdmccormick I renamed and moved all `calculate_progress ` methods to `lms/djangoapps/courseware/courses.py`;
2. @GlugovGrGlib will be able to answer the second question better. This decision was made taking into account the further development of the mobile API.

Reply:
@kdmccormick that's true this approach for the assignments completions and progress calculation is not similar to the one on progress page in LMS. This was a product decision made in collaboration with @marcotuts 
Please refer to [FIGMA designs](https://www.figma.com/design/iZ56YMjbRMShCCDxqrqRrR/Open-edX-Mobile-App-All-Screens-v2.1?node-id=9177-44208&t=meiXCsfDbe7B10Eq-0) and to the [product requirements write up](https://openedx.atlassian.net/wiki/spaces/OEPM/pages/3781754910/Project+-+Streamlined+Course+Navigation+Sequences+FC-0046#Course-Home) for more info.

We discussed that now data about assignments on progress page, and course home in web differs from mobile course home but in the future iterations the same feture may be applied to the web course home as well.